File: java/src/main/java/com/java/queue/LinkedQueue.java
package com.java.queue;

import com.java.list.Node;

public class LinkedQueue<E> implements QueueInterface<E> {
	private Node<E> tail;

	public LinkedQueue() {
		tail = null;
	}


	public static void main(String[] args) {
		LinkedQueue<String> q = new LinkedQueue<>();

		q.enqueue("hello");
		q.enqueue("java");
		q.enqueue("world");
		q.enqueue("!");


		System.out.println(q.dequeue());
		System.out.println(q.dequeue());

		System.out.println(q.isEmpty());

		System.out.println(q.dequeue());
		System.out.println(q.dequeue());

		System.out.println(q.isEmpty());




	}

	@Override
	public void enqueue(E x) {
		Node<E> newNode = new Node<>(x);
		if (isEmpty()) {
			newNode.next = newNode;
			tail = newNode;
		} else {
			newNode.next = tail.next;
			tail.next = newNode;
			tail = newNode;
		}

	}

	@Override
	public E dequeue() {
		if (isEmpty()) return null;
		else {
			Node<E> front = tail.next;
			if (front == tail) {
				tail = null;
			} else {
				tail.next = tail.next.next;
			}
			return front.item;
		}
	}

	@Override
	public E front() {
		if (isEmpty()) return null;
		else return tail.next.item;
	}

	@Override
	public boolean isEmpty() {
		return (tail == null);
	}

	@Override
	public void dequeueAll() {
		tail = null;
	}
}

File: kotlin/src/main/kotlin/com/kotlin/queue/LinkedQueue.kt
package com.kotlin.queue

import com.kotlin.list.Node

class LinkedQueue : QueueInterface {
    private var tail: Node? = null
    private lateinit var newNode: Node

    override fun enqueue(x: String) {
        newNode = Node(x, null)
        if (isEmpty()) {
            newNode.next = newNode
            tail = newNode
        } else {
            newNode.next = tail?.next
            tail?.next = newNode
            tail = newNode
        }
    }

    override fun dequeue(): String? {
        return if (isEmpty()) {
            null
        } else {
            val front = tail?.next
            if (front == tail) { // 하나일 때 뺏을 때
                tail = null
            } else {
                tail?.next = tail?.next?.next
            }
            front!!.item
        }
    }

    override fun front(): String? {
        return if (isEmpty()) {
            null
        } else {
            tail!!.next!!.item
        }
    }

    override fun isEmpty(): Boolean {
        return tail == null
    }

    override fun dequeueAll() {
        tail = null
    }






}